File: 36-valid-sudoku/valid-sudoku.cpp
// class Solution {
// public:
//     bool isValidSudoku(vector<vector<char>>& board) {
//         vector<unordered_set<int>> row(9);
//         vector<unordered_set<int>> col(9);
//         vector<vector<unordered_set<int>>> gridset(3,3);

//         for(int i=0;i<board.size();i++){
//             gridset.clear();
//             for(int j=0;j<board[0].size();j++){
//                 int num = board[i][j];
//                 if(num==".") continue;

//                 if(row[i].count(num) || col[j].count(num) || gridset[i][j%3].count(num)) return false;

//                 row[i].insert(num);
//                 col[j].insert(num);
//                 gridset[i][j%3].insert(num);
//             }
//         }
//         return true;
//     }
// };

#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<unordered_set<char>> row_sets(9);
        vector<unordered_set<char>> col_sets(9);
        unordered_set<char> grid_set;

        for (int i = 0; i < 9; ++i) {
            grid_set.clear();
            for (int j = 0; j < 9; ++j) {
                int row = i - i % 3 + j / 3;
                int col = (i % 3) * 3 + j % 3;
                char num = board[row][col];

                if (num == '.') continue;

                if (row_sets[row].count(num)) return false;
                if (col_sets[col].count(num)) return false;
                if (grid_set.count(num)) return false;

                row_sets[row].insert(num);
                col_sets[col].insert(num);
                grid_set.insert(num);
            }
        }
        return true;
    }
};

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_sets=[set() for _ in range(9)]
        col_sets=[set() for _ in range(9)]
        grid_set=set()
        for i in range(9):
            grid_set.clear()
            for j in range(3):
                for k in range(3):
                    num=board[j+(3*(i//3))][k+(3*(i%3))]
                    if num=='.':
                        continue
                    if num in row_sets[j+(3*(i//3))]:
                        return False
                    if num in col_sets[k+(3*(i%3))]:
                        return False
                    if num in grid_set:
                        return False
                    
                    row_sets[j+(3*(i//3))].add(num)
                    col_sets[k+(3*(i%3))].add(num)
                    grid_set.add(num)
        return True
        